crates/cheatcodes/src/test/expect.rs

```diff
@@ -698,72 +796,117 @@ pub(crate) fn handle_expect_revert(
         }
     };
 
-    ensure!(!matches!(status, return_ok!()), "next call did not revert as expected");
-
-    // If expected reverter address is set then check it matches the actual reverter.
-    if let (Some(expected_reverter), Some(actual_reverter)) =
-        (expected_revert.reverter, expected_revert.reverted_by)
-    {
-        if expected_reverter != actual_reverter {
-            return Err(fmt_err!(
-                "Reverter != expected reverter: {} != {}",
-                actual_reverter,
-                expected_reverter
-            ));
+    let stringify = |data: &[u8]| {
+        if let Ok(s) = String::abi_decode(data, true) {
+            return s;
         }
-    }
-
-    let expected_reason = expected_revert.reason.as_deref();
-    // If None, accept any revert.
-    let Some(expected_reason) = expected_reason else {
-        return Ok(success_return());
+        if data.is_ascii() {
+            return std::str::from_utf8(data).unwrap().to_owned();
+        }
+        hex::encode_prefixed(data)
     };
 
-    if !expected_reason.is_empty() && retdata.is_empty() {
-        bail!("call reverted as expected, but without data");
-    }
-
-    let mut actual_revert: Vec<u8> = retdata.into();
+    if expected_revert.count == 0 {
+        if expected_revert.reverter.is_none() && expected_revert.reason.is_none() {
+            ensure!(
+                matches!(status, return_ok!()),
+                "call reverted when it was expected not to revert"
+            );
+            return Ok(success_return());
+        }
 
-    // Compare only the first 4 bytes if partial match.
-    if expected_revert.partial_match && actual_revert.get(..4) == expected_reason.get(..4) {
-        return Ok(success_return())
-    }
+        // Flags to track if the reason and reverter match.
+        let mut reason_match = expected_revert.reason.as_ref().map(|_| false);
+        let mut reverter_match = expected_revert.reverter.as_ref().map(|_| false);
 
-    // Try decoding as known errors.
-    if matches!(
-        actual_revert.get(..4).map(|s| s.try_into().unwrap()),
-        Some(Vm::CheatcodeError::SELECTOR | alloy_sol_types::Revert::SELECTOR)
-    ) {
-        if let Ok(decoded) = Vec::<u8>::abi_decode(&actual_revert[4..], false) {
-            actual_revert = decoded;
+        // Reverter check
```

Comment:
this makes sense, cannot think of a cleaner way to structure code as we need to check both reverter and revert reason